### PR TITLE
Bypass (offloading)

### PR DESCRIPTION
**Pull request by glongo**
This is an updated patchset that implements bypass/shunting in NFQ.

Last PR: https://github.com/inliniac/suricata/pull/2113

- PR glongo: https://buildbot.openinfosecfoundation.org/builders/glongo/builds/121
- PR glongo-pcap: https://buildbot.openinfosecfoundation.org/builders/glongo-pcap/builds/120

### CODE REVIEW

File: suricata.yaml.in [hunk 1]
```diff
@@ -1103,6 +1103,7 @@ flow-timeouts:
 #   async-oneside: false        # don't enable async stream handling
 #   inline: no                  # stream inline mode
 #   max-synack-queued: 5        # Max different SYN/ACKs to queue
+#   offload: no                 # Enable offloading when stream.depth is reached
```
Comment (inliniac) [inline suricata.yaml.in]: s/offload/bypass/

File: src/decode.c [hunk 2]
```diff
@@ -382,6 +383,20 @@ void PacketDefragPktSetupParent(Packet *parent)
     DecodeSetNoPayloadInspectionFlag(parent);
 }
 
+int PacketHasBypassCallback(Packet *p)
```
Comment (inliniac) [inline src/decode.c]: why do you need this check function? Seems easier and more efficient to just call PacketBypassCallback and do the check there.

File: suricata.yaml.in [hunk 3]
```diff
@@ -1414,6 +1414,8 @@ nfq:
 #  mode: accept
 #  repeat-mark: 1
 #  repeat-mask: 1
+#  offload-mark: 1
```
Comment (inliniac) [inline suricata.yaml.in]: s/offload/bypass/g

File: src/source-nfq.c [hunk 4]
```diff
@@ -125,14 +125,15 @@ static int runmode_workers;
 typedef struct NFQThreadVars_
 {
     uint16_t nfq_index;
+    char *str_nfq_index;
```
Comment (inliniac) [inline src/source-nfq.c]: const?

File: src/util-device.h [hunk 5]
```diff
@@ -30,13 +30,15 @@ typedef struct LiveDevice_ {
...
     int ignore_checksum;
+    void (*BypassCallback)(Packet *p);
     SC_ATOMIC_DECLARE(uint64_t, pkts);
     SC_ATOMIC_DECLARE(uint64_t, drop);
     SC_ATOMIC_DECLARE(uint64_t, invalid_checksums);
     TAILQ_ENTRY(LiveDevice_) next;
 } LiveDevice;
 
 int LiveRegisterDevice(const char *dev, enum RunModes runmode);
+int LiveRegisterDeviceWithCallback(const char *dev, enum RunModes runmode, void (*BypassCallback)(Packet *p));
```
Comment (inliniac) [inline src/util-device.h]: LiveRegisterDeviceWithBypassCallback would be a more accurate name I think

File: src/detect-filestore.c [hunk 6]
```diff
@@ -421,3 +422,108 @@ static void DetectFilestoreFree(void *ptr)
...
+    StreamTcpInitConfig(TRUE);
+
+    de_ctx = DetectEngineCtxInit();
+    if (de_ctx == NULL)
+        goto end;
+
+    de_ctx->flags |= DE_QUIET;
+
+    de_ctx->sig_list = SigInit(de_ctx,"alert http any any -> any any "
+                               "(offload; filestore; "
```
Comment (inliniac) [inline src/detect-filestore.c]: offload

File: src/detect-filestore.c [hunk 7]
```diff
@@ -421,3 +422,108 @@ static void DetectFilestoreFree(void *ptr)
         SCFree(ptr);
     }
 }
+
+#ifdef UNITTESTS
+static int DetectFilestoreTest01(void)
```
Comment (inliniac) [inline src/detect-filestore.c]: can you add a description of what this test is doing?

File: src/detect-filestore.c [hunk 8]
```diff
@@ -292,6 +292,13 @@ static int DetectFilestoreSetup (DetectEngineCtx *de_ctx, Signature *s, char *st
     int ret = 0, res = 0;
     int ov[MAX_SUBSTRINGS];
 
+    /* filestore and offload keywords can't work together */
+    if (s->flags & SIG_FLAG_BYPASS) {
+        SCLogError(SC_ERR_CONFLICTING_RULE_KEYWORDS,
+                   "filestore can't work with offload keyword");
```
Comment (inliniac) [inline src/detect-filestore.c]: offload

File: src/stream-tcp.c [hunk 9]
```diff
@@ -4642,8 +4642,10 @@ int StreamTcpPacket (ThreadVars *tv, Packet *p, StreamTcpThread *stt,
             if (StreamTcpBypassEnabled() && PacketHasBypassCallback(p)) {
                 PacketBypassCallback(p);
             }
-        } else if ((ssn->client.flags & STREAMTCP_STREAM_FLAG_DEPTH_REACHED) ||
-                    (ssn->server.flags & STREAMTCP_STREAM_FLAG_DEPTH_REACHED))
+        }
+
+        if ((ssn->client.flags & STREAMTCP_STREAM_FLAG_DEPTH_REACHED) ||
```
Comment (inliniac) [inline src/stream-tcp.c]: This fix should go into the commit where the issue was introduced

File: src/stream-tcp.c [hunk 10]
```diff
@@ -4653,6 +4655,10 @@ int StreamTcpPacket (ThreadVars *tv, Packet *p, StreamTcpThread *stt,
             (PKT_IS_TOCLIENT(p) && (ssn->server.flags & STREAMTCP_STREAM_FLAG_NOREASSEMBLY)))
         {
             p->flags |= PKT_STREAM_NOPCAPLOG;
+            /* we can call bypass callback, if enabled */
+            if (StreamTcpBypassEnabled() && PacketHasBypassCallback(p)) {
```
Comment (inliniac) [inline src/stream-tcp.c]: the comment in suricata.yaml mentions only stream.depth, so it will need to be updated as well

File: src/decode.c (hunk 2 above)
Comment (glongo) [inline src/decode.c]: ack

File: suricata.yaml.in (hunk 1 above)
Comment (glongo) [inline suricata.yaml.in]: ack

File: src/detect-filestore.c (hunk 7 above)
Comment (glongo) [inline src/detect-filestore.c]: ack

File: src/detect-filestore.c (hunk 6 above)
Comment (glongo) [inline src/detect-filestore.c]: ack

File: src/detect-filestore.c (hunk 8 above)
Comment (glongo) [inline src/detect-filestore.c]: ack

File: suricata.yaml.in (hunk 3 above)
Comment (glongo) [inline suricata.yaml.in]: ack

File: src/source-nfq.c (hunk 4 above)
Comment (glongo) [inline src/source-nfq.c]: ack

File: src/stream-tcp.c (hunk 9 above)
Comment (glongo) [inline src/stream-tcp.c]: ack

**Pull request closed without merging**